File: fullstack/auth/paseto-cross-language/issuer-go/internal/token/paseto.go
// Package token signs and verifies PASETO v4.public access tokens.
//
// Interop contract shared with the Rust (pasetors) and TS (paseto-ts)
// verifiers: footer and implicit assertion are EMPTY everywhere, and claims
// are the RFC 3339 strings PASETO mandates (sub, role, iat, exp).
package token

import (
	"fmt"
	"time"

	"aidanwoods.dev/go-paseto"
)

// Claims is the verified content of an access token.
type Claims struct {
	Subject   string
	Role      string
	IssuedAt  time.Time
	ExpiresAt time.Time
}

// Signer holds the Ed25519 secret key. Only the issuer has one.
type Signer struct {
	secret paseto.V4AsymmetricSecretKey
}

func NewSigner(secretHex string) (*Signer, error) {
	secret, err := paseto.NewV4AsymmetricSecretKeyFromHex(secretHex)
	if err != nil {
		return nil, fmt.Errorf("token: bad secret key: %w", err)
	}
	return &Signer{secret: secret}, nil
}

// Sign issues a v4.public token. Returns the token and its expiry as the
// RFC 3339 string embedded in the exp claim.
func (s *Signer) Sign(subject, role string, now time.Time, ttl time.Duration) (string, string, error) {
	exp := now.Add(ttl)
	t := paseto.NewToken()
	t.SetSubject(subject)
	t.SetIssuedAt(now)
	// pasetors' default validation rules require nbf to be present; go-paseto
	// does not emit it on its own. Set it explicitly or Rust rejects the token.
	t.SetNotBefore(now)
	t.SetExpiration(exp)
	t.SetString("role", role)
	return t.V4Sign(s.secret, nil), exp.Format(time.RFC3339), nil
}

// Verifier holds only the Ed25519 public key.
type Verifier struct {
	public paseto.V4AsymmetricPublicKey
	parser paseto.Parser
}

func NewVerifier(publicHex string) (*Verifier, error) {
	public, err := paseto.NewV4AsymmetricPublicKeyFromHex(publicHex)
	if err != nil {
		return nil, fmt.Errorf("token: bad public key: %w", err)
	}
	// The default parser already enforces exp/iat/nbf validity.
	return &Verifier{public: public, parser: paseto.NewParser()}, nil
}

func (v *Verifier) Verify(raw string) (Claims, error) {
	t, err := v.parser.ParseV4Public(v.public, raw, nil)
	if err != nil {
		return Claims{}, fmt.Errorf("token: verify: %w", err)
	}
	subject, err := t.GetSubject()
	if err != nil {
		return Claims{}, fmt.Errorf("token: missing sub: %w", err)
	}
	role, err := t.GetString("role")
	if err != nil {
		return Claims{}, fmt.Errorf("token: missing role: %w", err)
	}
	issuedAt, err := t.GetIssuedAt()
	if err != nil {
		return Claims{}, fmt.Errorf("token: missing iat: %w", err)
	}
	expiresAt, err := t.GetExpiration()
	if err != nil {
		return Claims{}, fmt.Errorf("token: missing exp: %w", err)
	}
	return Claims{Subject: subject, Role: role, IssuedAt: issuedAt, ExpiresAt: expiresAt}, nil
}

File: fullstack/auth/paseto-cross-language/verifier-ts/src/paseto.ts
// PASETO v4.public verification with only the Ed25519 public key.
//
// Interop contract with the Go issuer and Rust verifier: footer and implicit
// assertion are EMPTY, claims are RFC 3339 strings (sub, role, iat, exp, nbf).
import { readFileSync } from "node:fs";
import { verify } from "paseto-ts/v4";

export interface Claims {
  subject: string;
  role: string;
  issuedAt: string;
  expiresAt: string;
}

// paseto-ts refuses raw key material: keys must be PASERK-scoped
// ("k4.public.<base64url>"). The other languages exchange bare hex, so wrap it.
export function hexToPaserkPublic(hex: string): string {
  const bytes = Buffer.from(hex.trim(), "hex");
  if (bytes.length !== 32) {
    throw new Error(`expected 32-byte Ed25519 public key, got ${bytes.length}`);
  }
  return `k4.public.${bytes.toString("base64url")}`;
}

export function devPublicKeyHex(): string {
  return readFileSync(new URL("../../keys/dev.public.hex", import.meta.url), "utf8").trim();
}

export async function verifyAccessToken(paserkPublicKey: string, raw: string): Promise<Claims> {
  // verify() checks the signature and validates exp/iat/nbf when present.
  const { payload } = await verify<{ sub?: string; role?: string; iat?: string; exp?: string }>(
    paserkPublicKey,
    raw,
  );
  const { sub, role, iat, exp } = payload;
  if (!sub || !role || !iat || !exp) {
    throw new Error("token is missing a required claim (sub, role, iat, exp)");
  }
  return { subject: sub, role, issuedAt: iat, expiresAt: exp };
}
